h2/src/main/org/h2/command/Command.java

```diff
@@ -238,74 +242,115 @@ public ResultInterface executeQuery(long maxrows, boolean scrollable) {
-                    } catch (DbException e) {
-                        // cannot retry some commands
-                        if (!isRetryable()) {
-                            throw e;
+            return executeUpdate(generatedKeysRequest, true);
+        } finally {
+            session.unlock();
+        }
+    }
+
+    @Override
+    public BatchResult executeBatchUpdate(ArrayList<Value[]> batchParameters, Object generatedKeysRequest) {
+        session.lock();
+        try {
+            session.waitIfExclusiveModeEnabled();
+            int size = batchParameters.size();
+            long[] updateCounts = new long[size];
+            MergedResult generatedKeys = generatedKeysRequest != null ? new MergedResult() : null;
+            ArrayList<SQLException> exceptions = new ArrayList<>();
+            for (int i = 0; i < size; i++) {
+                Value[] set = batchParameters.get(i);
+                ArrayList<? extends ParameterInterface> parameters = getParameters();
+                for (int j = 0, l = set.length; j < l; j++) {
+                    parameters.get(j).setValue(set[j], true);
+                }
+                long updateCount;
+                try {
+                    ResultWithGeneratedKeys result = executeUpdate(generatedKeysRequest, i + 1 == size);
+                    updateCount = result.getUpdateCount();
+                    if (generatedKeys != null) {
+                        ResultInterface keys = result.getGeneratedKeys();
+                        if (keys != null) {
+                            generatedKeys.add(keys);
                         }
-                        start = filterConcurrentUpdate(e, start);
-                    } catch (OutOfMemoryError e) {
-                        callStop = false;
-                        database.shutdownImmediately();
-                        throw DbException.convert(e);
-                    } catch (Throwable e) {
-                        throw DbException.convert(e);
                     }
+                } catch (Exception e) {
+                    exceptions.add(DbException.toSQLException(e));
+                    updateCount = Statement.EXECUTE_FAILED;
                 }
-            } catch (DbException e) {
-                e = e.addSQL(sql);
-                SQLException s = e.getSQLException();
-                database.exceptionThrown(s, sql);
-                if (s.getErrorCode() == ErrorCode.OUT_OF_MEMORY) {
-                    callStop = false;
-                    database.shutdownImmediately();
-                    throw e;
-                }
+                updateCounts[i] = updateCount;
+            }
+            return new BatchResult(updateCounts, generatedKeys != null ? generatedKeys.getResult() : null, exceptions);
+        } finally {
+            session.unlock();
+        }
+    }
+
+    private ResultWithGeneratedKeys executeUpdate(Object generatedKeysRequest, boolean commitIfAutoCommit) {
```

Comment:
Parameter `commitIfAutoCommit` is unused

Reply:
Good catch, I changed it temporarily and forgot to revert that change back. I've updated results of benchmark, insertions with many elements in the batch are now faster.